In [1]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression

In [2]:
data = pd.read_csv('https://raw.githubusercontent.com/joanby/python-ml-course/master/datasets/ecom-expense/Ecom%20Expense.csv')
data.head()

Transaction ID  Age    Items   Monthly Income  Transaction Time  Record  \
0         TXN001    42       10            7313        627.668127       5   
1         TXN002    24        8           17747        126.904567       3   
2         TXN003    47       11           22845        873.469701       2   
3         TXN004    50       11           18552        380.219428       7   
4         TXN005    60        2           14439        403.374223       2   

   Gender City Tier  Total Spend  
0  Female    Tier 1  4198.385084  
1  Female    Tier 2  4134.976648  
2    Male    Tier 2  5166.614455  
3  Female    Tier 1  7784.447676  
4  Female    Tier 2  3254.160485

In [3]:
dummy_gender = pd.get_dummies(data['Gender'], prefix = "Gender")
dummy_city = pd.get_dummies(data['City Tier'], prefix = "City")

In [4]:
dummy_gender.head()

Gender_Female  Gender_Male
0              1            0
1              1            0
2              0            1
3              1            0
4              1            0

In [5]:
dummy_city.head()

City_Tier 1  City_Tier 2  City_Tier 3
0            1            0            0
1            0            1            0
2            0            1            0
3            1            0            0
4            0            1            0

In [6]:
columns_names = data.columns.values.tolist()
columns_names

['Transaction ID',
 'Age ',
 ' Items ',
 'Monthly Income',
 'Transaction Time',
 'Record',
 'Gender',
 'City Tier',
 'Total Spend']

#### Aplicando un jon

In [7]:
data_new = data[columns_names].join(dummy_gender)
data_new.head()

Transaction ID  Age    Items   Monthly Income  Transaction Time  Record  \
0         TXN001    42       10            7313        627.668127       5   
1         TXN002    24        8           17747        126.904567       3   
2         TXN003    47       11           22845        873.469701       2   
3         TXN004    50       11           18552        380.219428       7   
4         TXN005    60        2           14439        403.374223       2   

   Gender City Tier  Total Spend  Gender_Female  Gender_Male  
0  Female    Tier 1  4198.385084              1            0  
1  Female    Tier 2  4134.976648              1            0  
2    Male    Tier 2  5166.614455              0            1  
3  Female    Tier 1  7784.447676              1            0  
4  Female    Tier 2  3254.160485              1            0

In [8]:
column_names = data_new.columns.values.tolist()
column_names

['Transaction ID',
 'Age ',
 ' Items ',
 'Monthly Income',
 'Transaction Time',
 'Record',
 'Gender',
 'City Tier',
 'Total Spend',
 'Gender_Female',
 'Gender_Male']

In [9]:
data_new = data_new[column_names].join(dummy_city)
data_new.head()

Transaction ID  Age    Items   Monthly Income  Transaction Time  Record  \
0         TXN001    42       10            7313        627.668127       5   
1         TXN002    24        8           17747        126.904567       3   
2         TXN003    47       11           22845        873.469701       2   
3         TXN004    50       11           18552        380.219428       7   
4         TXN005    60        2           14439        403.374223       2   

   Gender City Tier  Total Spend  Gender_Female  Gender_Male  City_Tier 1  \
0  Female    Tier 1  4198.385084              1            0            1   
1  Female    Tier 2  4134.976648              1            0            0   
2    Male    Tier 2  5166.614455              0            1            0   
3  Female    Tier 1  7784.447676              1            0            1   
4  Female    Tier 2  3254.160485              1            0            0   

   City_Tier 2  City_Tier 3  
0            0            0  
1            1            0  
2            1            0  
3            0            0  
4            1            0

In [10]:
# Estas columnas son acorde al tutorial
feature_cols = ['Monthly Income', 'Transaction Time', 'Gender_Female', 'Gender_Male', 'City_Tier 1', 'City_Tier 2', 'City_Tier 3']

In [11]:
X = data_new[feature_cols]
y = data_new['Total Spend']

In [12]:
lm = LinearRegression()
lm.fit(X,y)

LinearRegression()

In [13]:
# Ordenadas
print(lm.intercept_)
# Coeficientes
print(lm.coef_)

3655.729407690653
[   0.15297825    0.12372609  -94.15779883   94.15779883  119.6632516
  -16.67901801 -102.9842336 ]


In [14]:
list(zip(feature_cols, lm.coef_))

[('Monthly Income', 0.15297824609320518),
 ('Transaction Time', 0.12372608642620006),
 ('Gender_Female', -94.15779883032013),
 ('Gender_Male', 94.15779883032015),
 ('City_Tier 1', 119.66325160390097),
 ('City_Tier 2', -16.679018007990468),
 ('City_Tier 3', -102.98423359591051)]

In [15]:
# Modelo malito, ya que el r2 es bajo, debido a que falta informacion
lm.score(X,y)

0.19478920552885381

In [16]:
# Añadimos record
feature_cols = ['Monthly Income', 'Transaction Time', 'Gender_Female', 'Gender_Male', 'City_Tier 1', 'City_Tier 2', 'City_Tier 3','Record']
X = data_new[feature_cols]
y = data_new['Total Spend']

In [17]:
lm = LinearRegression()
lm.fit(X,y)

LinearRegression()

In [18]:
print(lm.intercept_)

-79.41713030137362


In [19]:
list(zip(feature_cols, lm.coef_))

[('Monthly Income', 0.14753898049205738),
 ('Transaction Time', 0.15494612549589545),
 ('Gender_Female', -131.02501325554567),
 ('Gender_Male', 131.0250132555456),
 ('City_Tier 1', 76.76432601049527),
 ('City_Tier 2', 55.138974309232474),
 ('City_Tier 3', -131.9033003197278),
 ('Record', 772.2334457445648)]

In [20]:
# El r2 mejoro un resto
lm.score(X,y)

0.9179923586131016

#### Calculamos la desviacion tipica de los residuos

- Modelo Global Total_Spend = -79.41713030137089 + 0.14753898049205738 * Monthly Income + 0.1549461254958966 * Transaction Time -131.02501325554584 * Gender_Female + 131.02501325554562 * Gender_Male + 76.76432601049518 * City_Tier 1 + 55.13897430923225 * City_Tier 2 -131.90330031972758 * City_Tier 3 + 772.2334457445643 * Record

- Reales scenario
    * Si es hombre y esta en city tier 1 =  -79.41713030137089 + 0.14753898049205738 * Monthly Income + 0.1549461254958966 * Transaction Time  + 131.02501325554562 * Gender_Male + 76.76432601049518 * City_Tier 1 + 772.2334457445643 * Record

    * Si es hombre y esta en city tier 2 = 
        
        -79.41713030137089 + 0.14753898049205738 * Monthly Income + 0.1549461254958966 * Transaction Time + 131.02501325554562 * Gender_Male  + 55.13897430923225 * City_Tier 2 + 772.2334457445643 * Record

    * Si es hombre y esta en city tier 3 = -79.41713030137089 + 0.14753898049205738 * Monthly Income + 0.1549461254958966 * Transaction Time -131.02501325554584 * Gender_Female + 131.02501325554562 * Gender_Male -131.90330031972758 * City_Tier 3 + 772.2334457445643 * Record

    * Si es mujer y esta en city tier 1 = -79.41713030137089 + 0.14753898049205738 * Monthly Income + 0.1549461254958966 * Transaction Time -131.02501325554584 * Gender_Female + 76.76432601049518 * City_Tier 1 + 772.2334457445643 * Record

    * Si es mujer y esta en city tier 2 = -79.41713030137089 + 0.14753898049205738 * Monthly Income + 0.1549461254958966 * Transaction Time -131.02501325554584 * Gender_Female + 55.13897430923225 * City_Tier 2 + 772.2334457445643 * Record

    * Si es mujer y esta en city tier 3 = -79.41713030137089 + 0.14753898049205738 * Monthly Income + 0.1549461254958966 * Transaction Time -131.02501325554584 * Gender_Female -131.90330031972758 * City_Tier 3 + 772.2334457445643 * Record

   

In [21]:
data_new['Predict'] = -79.41713030137089 + 0.14753898049205738 * data_new['Monthly Income'] + 0.1549461254958966 * data_new['Transaction Time'] -131.02501325554584 * data_new.Gender_Female + 131.02501325554562 * data_new.Gender_Male + 76.76432601049518 * data_new['City_Tier 1'] + 55.13897430923225 * data_new['City_Tier 2'] -131.90330031972758 * data_new['City_Tier 3'] + 772.2334457445643 * data_new['Record']

In [22]:
data_new.head()

Transaction ID  Age    Items   Monthly Income  Transaction Time  Record  \
0         TXN001    42       10            7313        627.668127       5   
1         TXN002    24        8           17747        126.904567       3   
2         TXN003    47       11           22845        873.469701       2   
3         TXN004    50       11           18552        380.219428       7   
4         TXN005    60        2           14439        403.374223       2   

   Gender City Tier  Total Spend  Gender_Female  Gender_Male  City_Tier 1  \
0  Female    Tier 1  4198.385084              1            0            1   
1  Female    Tier 2  4134.976648              1            0            0   
2    Male    Tier 2  5166.614455              0            1            0   
3  Female    Tier 1  7784.447676              1            0            1   
4  Female    Tier 2  3254.160485              1            0            0   

   City_Tier 2  City_Tier 3      Predict  
0            0            0  4903.696720  
1            1            0  4799.434826  
2            1            0  5157.082504  
3            0            0  8068.012996  
4            1            0  3581.980335

In [23]:
SSD = np.sum((data_new['Predict']-data_new['Total Spend'])**2)
SSD

1517733985.3408167

In [24]:
RSE = np.sqrt(SSD/(len(data_new) - len(feature_cols)-1))
RSE

803.1318809818166

In [25]:
sales_mean = np.mean(data_new['Total Spend'])
sales_mean

6163.176415976714

In [26]:
error = RSE/sales_mean
error

0.13031135680294162

In [27]:
data_new['Predict_2'] = lm.predict(X)

#### El tutor no divide, ni trabaja con datos de validación, seguimos el ejercicio por cuestiones academicas

In [28]:
data_new.head()

Transaction ID  Age    Items   Monthly Income  Transaction Time  Record  \
0         TXN001    42       10            7313        627.668127       5   
1         TXN002    24        8           17747        126.904567       3   
2         TXN003    47       11           22845        873.469701       2   
3         TXN004    50       11           18552        380.219428       7   
4         TXN005    60        2           14439        403.374223       2   

   Gender City Tier  Total Spend  Gender_Female  Gender_Male  City_Tier 1  \
0  Female    Tier 1  4198.385084              1            0            1   
1  Female    Tier 2  4134.976648              1            0            0   
2    Male    Tier 2  5166.614455              0            1            0   
3  Female    Tier 1  7784.447676              1            0            1   
4  Female    Tier 2  3254.160485              1            0            0   

   City_Tier 2  City_Tier 3      Predict    Predict_2  
0            0            0  4903.696720  4903.696720  
1            1            0  4799.434826  4799.434826  
2            1            0  5157.082504  5157.082504  
3            0            0  8068.012996  8068.012996  
4            1            0  3581.980335  3581.980335

#### Enmascarando variables dummies

In [29]:
dummy_gender = pd.get_dummies(data['Gender'], prefix = 'Gender').iloc[:,1:]
dummy_city = pd.get_dummies(data['City Tier'], prefix = 'City').iloc[:,1:]

In [30]:
data_new = data.join(dummy_gender)
data_new = data_new.join(dummy_city)

In [31]:
data_new.head()

Transaction ID  Age    Items   Monthly Income  Transaction Time  Record  \
0         TXN001    42       10            7313        627.668127       5   
1         TXN002    24        8           17747        126.904567       3   
2         TXN003    47       11           22845        873.469701       2   
3         TXN004    50       11           18552        380.219428       7   
4         TXN005    60        2           14439        403.374223       2   

   Gender City Tier  Total Spend  Gender_Male  City_Tier 2  City_Tier 3  
0  Female    Tier 1  4198.385084            0            0            0  
1  Female    Tier 2  4134.976648            0            1            0  
2    Male    Tier 2  5166.614455            1            1            0  
3  Female    Tier 1  7784.447676            0            0            0  
4  Female    Tier 2  3254.160485            0            1            0

In [32]:
feature_cols = ['Monthly Income', 'Transaction Time', 'Gender_Male', 'City_Tier 2', 'City_Tier 3','Record']
#Dividir set
X = data_new[feature_cols]
y = data_new['Total Spend']

#### El tutor no divide, ni trabaja con datos de validación, seguimos el ejercicio por cuestiones academicas

In [33]:
lm = LinearRegression()
lm = lm.fit(X,y)
y_predic = lm.predict(X)
y_predic

array([4903.69671986, 4799.4348257 , 5157.08250392, ..., 9050.6903581 ,
       8220.59609399, 6814.31308959])

In [34]:
lm.score(X,y)

0.9179923586131016

In [35]:
print(lm.intercept_)

-133.67781754642056


In [36]:
list(zip(feature_cols, lm.coef_))

[('Monthly Income', 0.14753898049205744),
 ('Transaction Time', 0.15494612549589631),
 ('Gender_Male', 262.05002651109595),
 ('City_Tier 2', -21.62535170126296),
 ('City_Tier 3', -208.66762633022324),
 ('Record', 772.2334457445635)]

- Los coeficientes anteriores se pueden comprender de la siguiente manera:
- La mujer que vive en city tier 2 y 3
- El hombre que vive en city tier 1

Coeficientes con todas las variables en el model
* ('Monthly Income', 0.1475389804920574),
* ('Transaction Time', 0.15494612549589393),
* ('Gender_Female', -131.02501325554675),
* ('Gender_Male', 131.02501325554664),
* ('City_Tier 1', 76.764326010495),
* ('City_Tier 2', 55.13897430923278),
* ('City_Tier 3', -131.90330031972783),
* ('Record', 772.2334457445645)

Los cambios se reflejan
* ('Monthly Income', 0.14753898049205744),
* ('Transaction Time', 0.15494612549589837),
* ('Gender_Male', 262.0500265110939),
* ('City_Tier 2', -21.62535170126288),
* ('City_Tier 3', -208.66762633022293),
* ('Record', 772.2334457445637)

Los cambios se reflejan en 
* Gender_Male: antes -> 131.02, despues -> 262.05 = (131.02 - (-131.02))
* Gender_Female: antes -> -131.02, despues -> 0
* City_Tier 1: antes ->  76.76, despues -> 0
* City_Tier 2: antes -> 55.13, despues -> -21.62 = (55.13 - 76.76)
* City_Tier 3: antes -> -131.90, despues -> -208.66 = (-131.90-76.76)